src/phpDebug.ts

```diff
@@ -312,8 +323,41 @@ class PhpDebugSession extends vscode.DebugSession {
                 })
                 server.listen(args.port || 9000, (error: NodeJS.ErrnoException) => (error ? reject(error) : resolve()))
             })
+
+        const createProxyClient = () =>
+            new Promise((resolve, reject) => {
+                const proxyClient = net.createConnection(
+                    {
+                        port: args.proxyPort || 9001,
+                        host: args.proxyHost,
+                    },
+                    () => {
+                        this.sendEvent(
+                            new vscode.OutputEvent(`DBGP proxy: connected to ${args.proxyHost} : ${args.proxyPort}\n`)
+                        )
+                        if (args.proxyHost && args.proxyPort) {
+                            this._proxyConnection = new xdebug.DbgpProxyConnection(
+                                proxyClient,
+                                args.ideKey || 'vsc',
+                                args.proxyPort,
+                                args.proxyHost,
+                                args.allowMultipleSessions || true
+                            )
+                        }
+                    }
+                )
+
+                proxyClient.on('end', () => {
+                    resolve()
+                })
+            })
+
         try {
             if (!args.noDebug) {
+                if (args.proxyHost) {
+                    await createProxyClient()
+                }
+
                 await createServer()
```

Comment:
So when using a proxy, we need both a proxy connection and a server connection? Could you explain why?

Reply:
**5.3** _[tldr](https://xdebug.org/docs-dbgp.php#initiating-a-debugging-session)_
IDE notify proxy of unique port and unique key
> To support proxies and JIT daemons, the IDE should be configured with a port pointing to the proxy/JIT. The IDE then makes a connection to the proxy when it starts ...
>
> The proxy listens for IDE connections on port 9001, and for debugger engine connections on port 9000.

When we read the above we interpreted it as a separate connection but I suppose that need not be the case.

Reply:
We wanted to implement it without having to touch much of the already established code. This has been successful. But ideally, when using a proxy, it should be the one to establish the connection as the normal xdebug connection won't get through. Thus, the code should be consolidated. yes?

Reply:
So the connection to the server won't be made unless the proxy is initialized. Proxy initialization couples the ip and ideKey, allowing the proxy to know where to send the data back.

The code could be improved by reusing the connection instead of having two.